File: android/app/src/main/java/coffee/crema/ui/ModeTargets.kt
package coffee.crema.ui

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import coffee.crema.core.MachineState
import coffee.crema.core.MmrRegister
import coffee.crema.core.ModeTargetInputs
import coffee.crema.core.ModeTargets
import coffee.crema.core.celsiusToFahrenheit
import coffee.crema.core.resolveModeTargets
import kotlinx.serialization.json.Json

// Service-mode (steam / hot water / flush) display targets — the values the
// Brew mode chips and the mode timers show. The resolution (machine
// ShotSettings/MMR value → the user's Quick-Controls value → legacy de1app
// default, with a partial payload's literal 0 treated as missing) lives in
// the core — `de1_domain::resolve_mode_targets`, review #41: this file and
// the web BrewDashboard's derivation had drifted. One shared call for the
// tablet BrewScreen and the phone PhoneBrewScreen.

/** Wire codec for the resolver FFI round-trip. */
private val modeJson = Json { ignoreUnknownKeys = true }

/** The legacy de1app defaults, should the bridge ever reject (never in
 *  practice) — matches the core's own fallback tier. */
private val MODE_TARGET_FALLBACK = ModeTargets(
    steamTempC = 148f, steamTimeoutS = 90f,
    hotWaterTempC = 92f, hotWaterVolumeMl = 250f, hotWaterTimeoutS = 30f,
    flushTempC = 95f, flushTimeS = 4f,
)

/** Resolve the service-mode targets from the UI snapshot via the core.
 *  The FlushTemp MMR word crosses raw (deci-°C) — the ÷10 lives core-side. */
fun modeTargets(ui: MainUiState): ModeTargets {
    val ss = ui.de1ShotSettings
    val inputs = ModeTargetInputs(
        machineSteamTempC = ss?.steam_temp,
        machineSteamTimeoutS = ss?.steam_timeout,
        machineHotWaterTempC = ss?.hot_water_temp,
        machineHotWaterVolumeMl = ss?.hot_water_volume,
        machineHotWaterTimeoutS = ss?.hot_water_timeout,
        machineFlushTempDeciC = ui.de1MachineInfo[MmrRegister.FlushTemp]?.toFloat(),
        qcSteamTempC = ui.qcSteamTempC,
        qcSteamTimeS = ui.qcSteamTimeS,
        qcHotWaterTempC = ui.qcHotWaterTempC,
        qcHotWaterVolumeMl = ui.qcHotWaterVolumeMl,
        qcFlushTempC = ui.qcFlushTempC,
        qcFlushTimeS = ui.qcFlushTimeS,
    )
    return runCatching {
        modeJson.decodeFromString(
            ModeTargets.serializer(),
            resolveModeTargets(modeJson.encodeToString(ModeTargetInputs.serializer(), inputs)),
        )
    }.getOrDefault(MODE_TARGET_FALLBACK)
}

/**
 * [modeTargets] memoized against its actual inputs, for composable call
 * sites — the Brew screens recompose on every telemetry tick, and the
 * resolver is an FFI + JSON round-trip that must not ride that hot path.
 */
@Composable
fun rememberModeTargets(ui: MainUiState): ModeTargets {
    val flushDeci = ui.de1MachineInfo[MmrRegister.FlushTemp]
    return remember(
        ui.de1ShotSettings, flushDeci, ui.qcSteamTempC, ui.qcSteamTimeS,
        ui.qcHotWaterTempC, ui.qcHotWaterVolumeMl, ui.qcFlushTempC, ui.qcFlushTimeS,
    ) { modeTargets(ui) }
}

/** The active service mode's display label, or null when no mode is running. */
fun modeLabel(state: MachineState?): String? = when (state) {
    MachineState.Steam -> "Steaming"
    MachineState.HotWater -> "Hot water"
    MachineState.HotWaterRinse -> "Flushing"
    else -> null
}

/** The running mode's target ceiling, seconds — the timeout the firmware
 *  enforces (the cap, not the typical session length); null when idle. */
fun modeTargetSeconds(state: MachineState?, t: ModeTargets): Float? = when (state) {
    MachineState.Steam -> t.steamTimeoutS
    MachineState.HotWater -> t.hotWaterTimeoutS
    MachineState.HotWaterRinse -> t.flushTimeS
    else -> null
}

/** Live `elapsed / total s` counter for an active mode's chip sub-label. */
fun modeRunningSub(elapsedMs: Long, targetS: Float): String =
    fmt("%.1f / %.0f s", elapsedMs / 1000f, targetS)

/** Compact whole-degree temperature for the dense mode chips, e.g. "148 °C". */
fun formatTempCompact(celsius: Float, unit: String): String =
    if (unit == "F") fmt("%.0f °F", celsiusToFahrenheit(celsius))
    else fmt("%.0f °C", celsius)

File: core/de1-domain/src/mode_targets.rs
//! Service-mode (steam / hot-water / flush) display-target resolution.
//!
//! The Brew screen's resting mode chips show the *target* values the
//! firmware will hold during the next session. Each field resolves with
//! the same precedence on every shell: **live machine value** (the
//! connect-time `ShotSettingsRead` snapshot / `FlushTemp` MMR read,
//! echoed on change) → **the user's persisted Quick-Controls dial** →
//! **legacy de1app default**. The machine value is what the firmware
//! currently has loaded; the QC value is the user's intent before the
//! read lands (or when nothing is connected).
//!
//! Previously each shell derived these independently (web
//! `BrewDashboard` `MODE_TARGET_SEC` / `MODE_TARGET_LABEL` /
//! `flushTempC`; Android `ModeTargets.kt`) and drifted — review #41
//! moves the one derivation here.

use typeshare::typeshare;

/// Legacy de1app default steam target temperature, °C.
pub const DEFAULT_STEAM_TEMP_C: f32 = 148.0;
/// Legacy de1app default steam timeout, seconds.
pub const DEFAULT_STEAM_TIMEOUT_S: f32 = 90.0;
/// Legacy de1app default hot-water target temperature, °C.
pub const DEFAULT_HOT_WATER_TEMP_C: f32 = 92.0;
/// Legacy de1app default hot-water volume, ml.
pub const DEFAULT_HOT_WATER_VOLUME_ML: f32 = 250.0;
/// Legacy de1app default hot-water timeout, seconds.
pub const DEFAULT_HOT_WATER_TIMEOUT_S: f32 = 30.0;
/// Legacy de1app default flush temperature, °C.
pub const DEFAULT_FLUSH_TEMP_C: f32 = 95.0;
/// Legacy 4 s flush window, seconds.
pub const DEFAULT_FLUSH_TIME_S: f32 = 4.0;

/// Everything the resolution consumes, all optional — a shell fills in
/// what it has. `machine_*` fields come from the DE1 itself (`None`
/// until the connect-time reads land); `qc_*` fields are the user's
/// Quick-Controls dials (`None` when never set). A literal `0` in any
/// field is treated as missing — a partial / pre-handshake
/// `ShotSettings` payload can carry zeros, which are meaningless as
/// targets.
#[typeshare]
#[derive(Debug, Clone, Copy, PartialEq, Default, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct ModeTargetInputs {
    /// Machine steam target temperature, °C (`ShotSettingsRead.steam_temp`).
    pub machine_steam_temp_c: Option<f32>,
    /// Machine steam timeout, seconds (`ShotSettingsRead.steam_timeout`).
    pub machine_steam_timeout_s: Option<f32>,
    /// Machine hot-water target temperature, °C (`ShotSettingsRead.hot_water_temp`).
    pub machine_hot_water_temp_c: Option<f32>,
    /// Machine hot-water volume, ml (`ShotSettingsRead.hot_water_volume`).
    pub machine_hot_water_volume_ml: Option<f32>,
    /// Machine hot-water timeout, seconds (`ShotSettingsRead.hot_water_timeout`).
    pub machine_hot_water_timeout_s: Option<f32>,
    /// The machine's stored flush setpoint — the raw `FlushTemp` MMR word
    /// (`0x00803844`), in **deci-°C** exactly as read off the wire; the
    /// conversion lives here, not in the shells.
    pub machine_flush_temp_deci_c: Option<f32>,
    /// Quick-Controls steam temperature, °C.
    pub qc_steam_temp_c: Option<f32>,
    /// Quick-Controls steam duration, seconds.
    pub qc_steam_time_s: Option<f32>,
    /// Quick-Controls hot-water temperature, °C.
    pub qc_hot_water_temp_c: Option<f32>,
    /// Quick-Controls hot-water volume, ml.
    pub qc_hot_water_volume_ml: Option<f32>,
    /// Quick-Controls flush temperature, °C.
    pub qc_flush_temp_c: Option<f32>,
    /// Quick-Controls flush duration, seconds.
    pub qc_flush_time_s: Option<f32>,
}

/// The resolved service-mode targets the Brew mode chips / banners show.
#[typeshare]
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModeTargets {
    /// Steam target temperature, °C.
    pub steam_temp_c: f32,
    /// Steam timeout, seconds — the firmware-enforced ceiling.
    pub steam_timeout_s: f32,
    /// Hot-water target temperature, °C.
    pub hot_water_temp_c: f32,
    /// Hot-water volume, ml.
    pub hot_water_volume_ml: f32,
    /// Hot-water timeout, seconds.
    pub hot_water_timeout_s: f32,
    /// Flush water temperature, °C.
    pub flush_temp_c: f32,
    /// Flush duration, seconds. No machine-side value exists — QC → default.
    pub flush_time_s: f32,
}

/// First positive finite value, else the fallback — `??` would happily
/// pass a partial payload's literal `0` through.
fn pos_or(v: Option<f32>, fallback: f32) -> f32 {
    v.filter(|x| x.is_finite() && *x > 0.0).unwrap_or(fallback)
}

/// Resolve the service-mode targets: machine → Quick Controls → legacy
/// default, field by field (see the module docs). The hot-water timeout
/// has no QC dial and the flush time has no machine value; each skips
/// the missing tier.
#[must_use]
pub fn resolve_mode_targets(inputs: &ModeTargetInputs) -> ModeTargets {
    ModeTargets {
        steam_temp_c: pos_or(
            inputs.machine_steam_temp_c,
            pos_or(inputs.qc_steam_temp_c, DEFAULT_STEAM_TEMP_C),
        ),
        steam_timeout_s: pos_or(
            inputs.machine_steam_timeout_s,
            pos_or(inputs.qc_steam_time_s, DEFAULT_STEAM_TIMEOUT_S),
        ),
        hot_water_temp_c: pos_or(
            inputs.machine_hot_water_temp_c,
            pos_or(inputs.qc_hot_water_temp_c, DEFAULT_HOT_WATER_TEMP_C),
        ),
        hot_water_volume_ml: pos_or(
            inputs.machine_hot_water_volume_ml,
            pos_or(inputs.qc_hot_water_volume_ml, DEFAULT_HOT_WATER_VOLUME_ML),
        ),
        hot_water_timeout_s: pos_or(
            inputs.machine_hot_water_timeout_s,
            DEFAULT_HOT_WATER_TIMEOUT_S,
        ),
        flush_temp_c: pos_or(
            inputs.machine_flush_temp_deci_c.map(|d| d / 10.0),
            pos_or(inputs.qc_flush_temp_c, DEFAULT_FLUSH_TEMP_C),
        ),
        flush_time_s: pos_or(inputs.qc_flush_time_s, DEFAULT_FLUSH_TIME_S),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_inputs_resolve_to_the_legacy_defaults() {
        let t = resolve_mode_targets(&ModeTargetInputs::default());
        assert_eq!(t.steam_temp_c, DEFAULT_STEAM_TEMP_C);
        assert_eq!(t.steam_timeout_s, DEFAULT_STEAM_TIMEOUT_S);
        assert_eq!(t.hot_water_temp_c, DEFAULT_HOT_WATER_TEMP_C);
        assert_eq!(t.hot_water_volume_ml, DEFAULT_HOT_WATER_VOLUME_ML);
        assert_eq!(t.hot_water_timeout_s, DEFAULT_HOT_WATER_TIMEOUT_S);
        assert_eq!(t.flush_temp_c, DEFAULT_FLUSH_TEMP_C);
        assert_eq!(t.flush_time_s, DEFAULT_FLUSH_TIME_S);
    }

    #[test]
    fn the_machine_value_outranks_the_qc_dial() {
        let t = resolve_mode_targets(&ModeTargetInputs {
            machine_steam_temp_c: Some(160.0),
            qc_steam_temp_c: Some(140.0),
            ..ModeTargetInputs::default()
        });
        assert_eq!(t.steam_temp_c, 160.0);
    }

    #[test]
    fn a_partial_payloads_zero_falls_through_to_qc_then_default() {
        let t = resolve_mode_targets(&ModeTargetInputs {
            machine_steam_temp_c: Some(0.0),
            qc_steam_temp_c: Some(140.0),
            machine_hot_water_temp_c: Some(0.0),
            ..ModeTargetInputs::default()
        });
        assert_eq!(t.steam_temp_c, 140.0);
        assert_eq!(t.hot_water_temp_c, DEFAULT_HOT_WATER_TEMP_C);
    }

    #[test]
    fn the_flush_mmr_word_is_deci_celsius() {
        let t = resolve_mode_targets(&ModeTargetInputs {
            machine_flush_temp_deci_c: Some(883.0),
            qc_flush_temp_c: Some(90.0),
            ..ModeTargetInputs::default()
        });
        assert!((t.flush_temp_c - 88.3).abs() < 1e-4);
    }

    #[test]
    fn inputs_round_trip_the_camel_case_wire() {
        // The shells build this JSON — pin one field of each casing family.
        let parsed: ModeTargetInputs =
            serde_json::from_str(r#"{"machineSteamTempC":155,"qcFlushTimeS":6}"#).unwrap();
        assert_eq!(parsed.machine_steam_temp_c, Some(155.0));
        assert_eq!(parsed.qc_flush_time_s, Some(6.0));
        let t = resolve_mode_targets(&parsed);
        assert_eq!(t.steam_temp_c, 155.0);
        assert_eq!(t.flush_time_s, 6.0);
    }
}
